device/cuda/include/traccc/cuda/clusterization/clusterization_algorithm.hpp

```diff
@@ -67,12 +78,31 @@ class clusterization_algorithm
     /// @param det_descr The detector description
     /// @return a measurement collection (buffer)
     ///
-    output_type operator()(
+    /// @{
+    measurement_collection_types::buffer operator()(
         const edm::silicon_cell_collection::const_view& cells,
         const silicon_detector_description::const_view& det_descr)
         const override;
 
+    measurement_collection_types::buffer operator()(
+        const edm::silicon_cell_collection::const_view& cells,
+        const silicon_detector_description::const_view& det_descr,
+        device::clustering_discard_disjoint_set&&) const override;
+
+    std::pair<measurement_collection_types::buffer,
+              vecmem::data::vector_buffer<unsigned int>>
+    operator()(const edm::silicon_cell_collection::const_view& cells,
+               const silicon_detector_description::const_view& det_descr,
+               device::clustering_keep_disjoint_set&&) const override;
+    /// @}
+
     private:
+    std::pair<measurement_collection_types::buffer,
+              std::optional<vecmem::data::vector_buffer<unsigned int>>>
+    execute_impl(const edm::silicon_cell_collection::const_view& cells,
+                 const silicon_detector_description::const_view& det_descr,
+                 bool keep_disjoint_set) const;
```

Comment:
`std::optional` is a possibility, but wouldn't it be easier to have the following interface for this internal function instead?

```c++
void execute_impl(const edm::silicon_cell_collection::const_view& cells,
                  const silicon_detector_description::const_view& det_descr,
                  measurement_collection_types::buffer& meas,
                  edm::silicon_cluster_collection::buffer* clusters = nullptr) const;
```

Though maybe not... 🤔 Still, something to think about a little.

Reply:
No, I really do not want to add output arguments unless absolutely necessary!